src/txn/txn.c

```diff
@@ -675,87 +652,134 @@ __txn_fixup_prepared_update(WT_SESSION_IMPL *session, WT_TXN_OP *op, WT_CURSOR *
-            upd->next = cbt->ins->upd;
-        else if (cbt->ref->page->modify != NULL && cbt->ref->page->modify->mod_row_update != NULL)
-            upd->next = cbt->ref->page->modify->mod_row_update[cbt->slot];
-        WT_ASSERT(session, upd->next != NULL && upd->next->txnid == WT_TXN_ABORTED);
-
-        /* Append a tombstone if the stop timestamp exists. */
-        if (hs_stop_ts != WT_TS_MAX) {
-            WT_ERR(__wt_upd_alloc(session, NULL, WT_UPDATE_TOMBSTONE, &tombstone, NULL));
-            tombstone->durable_ts = hs_stop_ts;
-            tombstone->start_ts = hs_stop_ts;
-            tombstone->txnid = WT_TXN_NONE;
-            tombstone->next = upd;
-        } else
-            tombstone = upd;
+    /* Walk to the end of the chain. */
+    for (; chain->next != NULL; chain = chain->next)
+        ;
+
+    /* Append the update to the end of the chain. */
+    WT_PUBLISH(chain->next, tombstone);
+    *upd_appended = true;
+
+    __wt_cache_page_inmem_incr(session, page, total_size);
+
+    if (0) {
+err:
+        __wt_free(session, upd);
+        __wt_free(session, tombstone);
+    }
+done:
+    __wt_scr_free(session, &hs_key);
+    __wt_scr_free(session, &hs_value);
+    return (ret);
+}
+
+/*
+ * __txn_fixup_prepared_update --
+ *     Fix/restore the history store update of a prepared datastore update based on transaction
+ *     status.
+ */
+static int
+__txn_fixup_prepared_update(
+  WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_UPDATE *fix_upd, bool commit)
+{
+    WT_CURSOR_BTREE *hs_cbt;
+    WT_DECL_RET;
+    WT_ITEM hs_value;
+    WT_TXN *txn;
+    WT_UPDATE *hs_upd;
+    uint32_t txn_flags;
+
+    hs_cbt = (WT_CURSOR_BTREE *)hs_cursor;
+    txn = session->txn;
+
+    /*
+     * Transaction error and prepare are cleared temporarily as cursor functions are not allowed
+     * after an error or a prepared transaction.
+     */
+    txn_flags = FLD_MASK(txn->flags, WT_TXN_ERROR | WT_TXN_PREPARE);
+    F_CLR(txn, txn_flags);
 
-        WT_WITH_BTREE(session, CUR2BT(cbt),
-          ret = __wt_row_modify(cbt, &cbt->iface.key, NULL, tombstone, WT_UPDATE_INVALID, true));
-        WT_ERR(ret);
-        tombstone = NULL;
-        upd = NULL;
+    /* The value older than the prepared update in the history store must be a full value. */
+    WT_ASSERT(session, fix_upd->type == WT_UPDATE_STANDARD);
+
+    /*
+     * If we found a history value that satisfied the given timestamp, add it to the update list.
+     * Otherwise remove the key by adding a tombstone.
+     */
+    if (commit) {
+        WT_ERR(__wt_upd_alloc_tombstone(session, &hs_upd, NULL));
+        hs_upd->start_ts = txn->commit_timestamp;
+        hs_upd->durable_ts = txn->durable_timestamp;
+        hs_upd->txnid = txn->id;
 
+        hs_value.data = fix_upd->data;
+        hs_value.size = fix_upd->size;
+        hs_cursor->set_value(hs_cursor, txn->durable_timestamp, fix_upd->durable_ts,
+          (uint64_t)fix_upd->type, &hs_value);
```

Comment:
Do we use this set_value/upd_alloc pattern elsewhere? I haven't seen it, and it seems a bit odd to me.

Reply:
I think this is the only place.

Reply:
Add a comment why we do this

Reply:
done